BedrockServer.cpp

```diff
@@ -288,41 +288,61 @@ void BedrockServer::sync(SData& args,
+
+                        // This command completed in peek, respond to it appropriately, either directly or by sending it
+                        // back to the sync thread.
+                        SASSERT(command.complete);
+                        if (command.initiatingPeerID) {
+                            command.finalizeTimingInfo();
+                            syncNode.sendResponse(command);
+                        } else {
+                            server._reply(command);
+                        }
+                        continue;
                     }
-                    continue;
+                } else {
+                    // If the command had an https request before we tried to peek it, we record that, as we'll need to
+                    // restart the transaction.
+                    alreadyHadHttpsRequest = true;
                 }
-            }
 
-            // If we've dequeued a command with an incomplete HTTPS request, we move it to httpsCommands so that every
-            // subsequent dequeue doesn't have to iterate past it while ignoring it. Then we'll just start on the next
-            // command.
-            if (command.httpsRequest && !command.httpsRequest->response) {
-                httpsCommands.push_back(move(command));
-                continue;
-            }
+                // If we've dequeued a command with an incomplete HTTPS request, we move it to httpsCommands so that every
+                // subsequent dequeue doesn't have to iterate past it while ignoring it. Then we'll just start on the next
+                // command.
+                if (command.httpsRequest && !command.httpsRequest->response) {
+                    // We can't finish this transaction right now, so we roll it back. We'll restart it later when the
+                    // httpsRequest is complete.
+                    core.rollback();
 
-            // And now we'll decide how to handle it.
-            if (nodeState == SQLiteNode::MASTERING) {
-                // Now that we've peeked without finishing the command, we grab the commit mutex exclusively, so that none
-                // of our worker threads can attempt to write to the database until we're finished.
-                server._syncThreadCommitMutex.lock();
-                if (core.processCommand(command)) {
+                    // Done with the lock.
+                    server._syncThreadCommitMutex.unlock();
+
+                    // Set this aside and move on to the next command.
+                    httpsCommands.push_back(move(command));
+                    continue;
+                }
+
+                // if alreadyHadHttpsRequest is set, then we peeked this request at some point in the past and ended
+                // that transaction, so we need to start a new one, so we pass that as the `beginTransaction` flag.
+                if (core.processCommand(command, alreadyHadHttpsRequest)) {
```

Comment:
I'm confused why we need `alreadyHadHttpsRequest` at all.  Can't we just use `command.httpsRequest`?  In fact, don't we *need* to else any `HTTPSRequest` started by the previous peek would get missed... hm.  I still think all this would be more obvious if we just made `SQLite` keep track of whether or not it has an active transaction.  Then we're making decisions based on the actual thing we care of (eg, "is there an active transaction") rather than based on something that only infers that.